### DISCUSSION TITLE
About continuou_dims

### DISCUSSION THREAD

**Discussion by sangfeel** [TheStageAI/TorchIntegral#18] [Q&A] [answered]
continuous_dims = {
"layer4.0.conv1.weight": [0, 1],
"layer4.0.conv2.weight": [0, 1],
"layer4.0.downsample.0.weight": [0, 1],
"layer4.1.conv1.weight": [0, 1],
"layer4.1.conv2.weight": [0, 1],
}

Above is the definition of continuous_dim in imagnet.py .

The question is, what are the criteria for determining continuous_dim. Described as continuous_dim performing the crop. If so, do I have to choose all the parts that correspond to the last layer before entering the fully connected layer?

Additionally, in the example above, that's how it works
import torch
Bringing a torch to an inn
Get .reset back from torchvision 18

Model = resnet18 (pre-training = true)
Wrapping paper = Inn.All-in-one wrapper (init_from_discrete=True)

Specify continuous dimensions to cut
Continuous_dims = {
"Tier 4.0.conv1.weight": [0],
"Tier 4.1.conv1"Weight": [0, 1]
}

Converting to an integrated model
in_model = rapper (model, (1, 3, 224, 224), continuous_dims)

The choice of continuous_dims is slightly different here. I would appreciate it if you could tell me the difference between these two as well.

**Comment by b1n0** [comment, marked as answer]
In this dict you should specify parameter names and axes of corresponding tensors which you want to make continuous for further pruning. Thus, if you want to prune convolution _layer4.0.conv1_ along output dimension you should add _{'layer4.0.conv1.weight': [0]}_ to the continuous_dims dict. If input channels dimension of the this convolution layer is also pruned, then _{'layer4.0.conv1.weight': [0, 1]}_ is used.
**Note** that _bias_ parameter of convolution layer must have the same sampling grid as _weight_ parameter along out channels, but it is not necessary to list biases in continuous_dims, because torch_integral.graph module automatically detects such related tensors.

If the question is why do we prune these layers, it is just heuristic that last layers is easier to prune.

If you want to reproduce imagenet experiments use definition of continuous_dim in imagnet.py, else you can choose any layers you want.
